cpp/utils.cpp

```diff
@@ -207,6 +207,59 @@ double dolfinx_contact::dR_minus(double x) { return double(x < 0); }
 //-------------------------------------------------------------------------------------
 
 double dolfinx_contact::dR_plus(double x) { return double(x > 0); }
+
+std::array<double, 3> dolfinx_contact::ball_projection(std::array<double, 3> x,
+                                                       double alpha,
+                                                       std::size_t bs)
+{
+  // Compute norm of vector
+  double norm = 0;
+  for (std::size_t j = 0; j < bs; ++j)
+    norm += x[j] * x[j];
```

Comment:
If the third coordinate is 0 (and not unset, like infinity), it might be faster to do fixed loops of size 3, or using `for_each`, like
```suggestion
  double norm = 0;
  std::for_each(x.cbegin(), x.cend(),
                    [&norm](auto e) { norm += std::pow(e, 2); });
```